lib/ansible/module_utils/ecs/api.py

```diff
@@ -0,0 +1,364 @@
+            if req.getcode() != 429:
+                break
+            else:
+                try:
+                    retry_after = int(dict(req.headers).get("Retry-After")) + 1
+                except ValueError:
+                    retry_after = 10
+                time.sleep(retry_after)
+
+        # Return the result if JSON and success ({} for empty responses)
+        # Raise an exception if there was a failure.
+        try:
+            result = json.loads(req.read())
+        except ValueError:
+            result = {}
+
+        if req.getcode() >= 400:
+            request_error = True
+        else:
+            request_error = False
+
+        if result or result == {}:
+            if not request_error:
+                return result
+            else:
+                raise RestOperationException(result)
+
+        # Raise a generic RestOperationException if this fails
+        raise RestOperationException({"status": req.getcode(), "errors": [{"message": "REST Operation Failed"}]})
+
+
+class Resource(object):
+    """ Implement basic CRUD operations against a path. """
+
+    def __init__(self, session):
+        self.session = session
+        self.parameters = {}
+
+        for url in session._spec.get("paths").keys():
+            methods = session._spec.get("paths").get(url)
+            for method in methods.keys():
+                operation_spec = methods.get(method)
+                operation_name = operation_spec.get("operationId", None)
+                parameters = operation_spec.get("parameters")
+
+                if not operation_name:
+                    if method.lower() == "post":
+                        operation_name = "Create"
+                    elif method.lower() == "get":
+                        operation_name = "Get"
+                    elif method.lower() == "put":
+                        operation_name = "Update"
+                    elif method.lower() == "delete":
+                        operation_name = "Delete"
+                    elif method.lower() == "patch":
+                        operation_name = "Patch"
+                    else:
+                        raise SessionConfigurationException(to_native("Invalid REST method type {0}".format(method)))
+
+                    # Get the non-parameter parts of the URL and append to the operation name
+                    # e.g  /application/version -> GetApplicationVersion
+                    # e.g. /application/{id}    -> GetApplication
+                    # This may lead to duplicates, which we must prevent.
+                    operation_name += re.sub(r"{(.*)}", "", url).replace("/", " ").title().replace(" ", "")
+                    operation_spec["operationId"] = operation_name
+
+                op = RestOperation(session, url, method, parameters)
+                setattr(self, operation_name, bind(self, op.restmethod, operation_spec))
+
+
+# Functions similarly to requests.session, except leveraging the ansible urls package
```

Comment:
Is there a reason you are not using `Request` from https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/urls.py#L1031 ?

Reply:
I am. The Session object contains an instantiation of "Request" with various values configured
45: 'from ansible.module_utils.urls import Request'
247: `self.request = Request(headers=headers, timeout=60)'

The api.py Session had at one time been an extension of the "requests package" session's class, my comment was to remind me that module_util.urls.Requests is essentially similar to "Session". I can see how it's misleading and will remove.